msal_extensions/_windows.py

```diff
@@ -0,0 +1,173 @@
+import os
+import ctypes
+from ctypes import wintypes
+import time
+import msal
+import errno
+from .cache_lock import CrossPlatLock
+
+_local_free = ctypes.windll.kernel32.LocalFree
+_memcpy = ctypes.cdll.msvcrt.memcpy
+_crypt_protect_data = ctypes.windll.crypt32.CryptProtectData
+_crypt_unprotect_data = ctypes.windll.crypt32.CryptUnprotectData
+_CRYPTPROTECT_UI_FORBIDDEN = 0x01
+
+
+class DATA_BLOB(ctypes.Structure):
+    """
+    A wrapper for interacting with the _CRYPTOAPI_BLOB type and its many aliases. This type is exposed from Wincrypt.h
+    in XP and above.
+
+    See documentation for this type at:
+    https://msdn.microsoft.com/en-us/7a06eae5-96d8-4ece-98cb-cf0710d2ddbd
+    """
+    _fields_ = [("cbData", wintypes.DWORD), ("pbData", ctypes.POINTER(ctypes.c_char))]
+
+    def raw(self):
+        # type: () -> bytes
+        cb_data = int(self.cbData)
+        pb_data = self.pbData
+        buffer = ctypes.c_buffer(cb_data)
+        _memcpy(buffer, pb_data, cb_data)
+        _local_free(pb_data)
```

Comment:
So, the local_free is definitely necessary as we don't want to create a memory leak! However, you're totally right that it should be safe to call this method twice in a way it just isn't right now. I'll fix that up.

Reply:
Fixed in b9becbcd8

Reply:
So, I'm going to have to walk this change back. I got to a Windows machine and ran the tests, and curiously pulling this object out into the destructor causes a seg-fault that crashes the host Python process with it.

Reply:
If that is the case, can we have one more try as to move that `_local_free()` out of `raw()`, and put it into `protect()` in this manner? On paper, it should work.

```python
    def protect(self, ...):
        ...
        try:
            return result.raw()
        finally:
            _local_free(result.pbData)
```

Reply:
Love it :) I'll give it another whack in the morning.

Reply:
So the trick ended up being that LocalFree **MUST NOT** be called memory allocated by ctypes.create_string_buffer, but MUST be called on memory allocated by the CryptProtectData and CryptUnprotectData functions. It just so happened that I only ever called `raw` on the latter, and never called `raw` for the former. This means that using `try` & `finally` allows us to manage the lifetime of only specific structs, where as moving it into a `__del__` method was breaking the contract.

Long story short, this is fixed in  161e6cf. On the otherhand, a4efa654 can be ignored.

Reply:
Thanks for doing all the hard work to validate our latest fix, and documeting your finding! For the record, even if that 80e735c would work, I would still prefer the 2a25b5f approach. Because that's fundamentally the right pattern to manage memory (came from my 7 years on C/C++).

🚢